translator/translate/otel/pipeline/jmx/translator.go

```diff
@@ -0,0 +1,105 @@
+	"github.com/aws/amazon-cloudwatch-agent/translator/translate/otel/processor/ec2taggerprocessor"
+	"github.com/aws/amazon-cloudwatch-agent/translator/translate/otel/processor/filterprocessor"
+	"github.com/aws/amazon-cloudwatch-agent/translator/translate/otel/processor/resourceprocessor"
+	"github.com/aws/amazon-cloudwatch-agent/translator/translate/otel/receiver/jmx"
+)
+
+const (
+	placeholderTarget = "<type>"
+)
+
+type translator struct {
+	name  string
+	index int
+}
+
+type Option func(any)
+
+func WithIndex(index int) Option {
+	return func(a any) {
+		if t, ok := a.(*translator); ok {
+			t.index = index
+		}
+	}
+}
+
+var _ common.Translator[*common.ComponentTranslators] = (*translator)(nil)
+
+func NewTranslator(opts ...Option) common.Translator[*common.ComponentTranslators] {
+	t := &translator{name: common.PipelineNameJmx, index: -1}
+	for _, opt := range opts {
+		opt(t)
+	}
+	if t.index != -1 {
+		t.name += "/" + strconv.Itoa(t.index)
+	}
+	return t
+}
+
+func (t *translator) ID() component.ID {
+	return component.NewIDWithName(component.DataTypeMetrics, t.name)
+}
+
+// Translate creates a pipeline for jmx if jmx metrics are collected
+// section is present.
+func (t *translator) Translate(conf *confmap.Conf) (*common.ComponentTranslators, error) {
+	if conf == nil || !conf.IsSet(common.JmxConfigKey) {
+		return nil, &common.MissingKeyError{ID: t.ID(), JsonKey: common.JmxConfigKey}
+	}
+
+	if !hasMeasurements(conf, t.index) {
+		baseKey := common.JmxConfigKey
+		if t.index != -1 {
+			baseKey = fmt.Sprintf("%s[%d]", baseKey, t.index)
+		}
+		return nil, &common.MissingKeyError{ID: t.ID(), JsonKey: common.ConfigKey(baseKey, placeholderTarget, common.MeasurementKey)}
+	}
+
+	translators := common.ComponentTranslators{
+		Receivers: common.NewTranslatorMap(jmx.NewTranslator(jmx.WithIndex(t.index))),
+		Processors: common.NewTranslatorMap(
+			filterprocessor.NewTranslator(filterprocessor.WithName(common.PipelineNameJmx), filterprocessor.WithIndex(t.index)),
+			resourceprocessor.NewTranslator(resourceprocessor.WithName(common.PipelineNameJmx)),
+		),
+		Exporters:  common.NewTranslatorMap(awscloudwatch.NewTranslator()),
+		Extensions: common.NewTranslatorMap(agenthealth.NewTranslator(component.DataTypeMetrics, []string{agenthealth.OperationPutMetricData})),
+	}
+
+	if conf.IsSet(common.ConfigKey(common.MetricsKey, common.AppendDimensionsKey)) {
+		translators.Processors.Set(ec2taggerprocessor.NewTranslator())
+	}
+
+	return &translators, nil
+}
+
+func hasMeasurements(conf *confmap.Conf, index int) bool {
+	jmxMap := common.GetJmxMap(conf, index)
+	if len(jmxMap) == 0 {
+		return false
+	}
+	var validTargets []string
+	for _, target := range common.JmxTargets {
+		if targetMap, ok := jmxMap[target].(map[string]any); ok {
+			if measurements, ok := targetMap[common.MeasurementKey].([]any); !ok || len(measurements) == 0 {
+				return false
+			}
+			validTargets = append(validTargets, target)
+		}
+	}
+	return len(validTargets) > 0
```

Comment:
We do because if we don't ensure that we have `validTargets`, then

```json
{
  "jmx": {
    "jvm": {
    }
  }
}
```

or even


```json
{
  "jmx": {
  }
}
```
would be considered valid.

It's kinda covered in the `TestTranslator/WithEmpty` test case, but I can add another.

Reply:
Gotcha... could it then be a boolean and set that to true where you do the append currently?
Just to save on space for an array.

Reply:
Hmm yeah you're right. I think at some point I was going to do something else with it. I'll fix it.